sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -144,6 +144,15 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
           value != null =>
       Some(unwrapTimestampToDate(be, fromExp, ts, timeZoneId, evalMode))
 
+    // Timestamp/Timestamp_NTZ -> Timestamp_NTZ/Timestamp
```

Comment:
`cast(ts_ntz_col as ts) compare ts_lit` means `convertTz(col_value, tz, UTC) compare lit_value`, and we are trying to reverse it to `col_value compare convertTz(lit_value, UTC, tz)`.

We need to be careful here, as datetime with timezone is very tricky. My proposal is to do a round trip tests before rewriting it: if `convertTz(convertTz(lit_value, UTC, tz), tz, UTC) == lit_value`. This means we convert the comparison to `convertTz(col_value, tz, UTC) compare convertTz(convertTz(lit_value, UTC, tz), tz, UTC) == lit_value` first, and we strip the `convertTz(_, tz, UTC)` from both sides.

Reply:
Added round trip check before the rewriting it.

Reply:
I use `Cast.eval` instead of `convertTz` directly because otherwise we need to handle timestamp -> timestamp_ntz and timestamp_ntz -> timestamp two cases here. Let me know if you prefer we use `convertTz` directly here.